fish-rust/src/input_common.rs

```diff
@@ -1,21 +1,586 @@
+    let mut fdset = FdReadableSet::new();
+    loop {
+        fdset.clear();
+        fdset.add(in_fd);
+
+        // Add the completion ioport.
+        let ioport_fd = iothread_port();
+        fdset.add(ioport_fd);
+
+        // Get the uvar notifier fd (possibly none).
+        let notifier = default_notifier();
+        let notifier_fd = notifier.notification_fd();
+        if let Some(notifier_fd) = notifier.notification_fd() {
+            fdset.add(notifier_fd);
+        }
+
+        // Here's where we call select().
+        let select_res = fdset.check_readable(FdReadableSet::kNoTimeout);
+        if select_res < 0 {
+            let err = errno::errno().0;
+            if err == libc::EINTR || err == libc::EAGAIN {
+                // A signal.
+                return ReadbResult::Interrupted;
+            } else {
+                // Some fd was invalid, so probably the tty has been closed.
+                return ReadbResult::Eof;
+            }
+        }
+
+        // select() did not return an error, so we may have a readable fd.
+        // The priority order is: uvars, stdin, ioport.
+        // Check to see if we want a universal variable barrier.
+        if let Some(notifier_fd) = notifier_fd {
+            if fdset.test(notifier_fd) && notifier.notification_fd_became_readable(notifier_fd) {
+                return ReadbResult::UvarNotified;
+            }
+        }
+
+        // Check stdin.
+        if fdset.test(in_fd) {
+            let mut arr: [u8; 1] = [0];
+            if read_blocked(in_fd, &mut arr) != 1 {
+                // The terminal has been closed.
+                return ReadbResult::Eof;
+            }
+            // The common path is to return a u8.
+            return ReadbResult::Byte(arr[0]);
+        }
+
+        // Check for iothread completions only if there is no data to be read from the stdin.
+        // This gives priority to the foreground.
+        if fdset.test(ioport_fd) {
+            return ReadbResult::IOPortNotified;
+        }
+    }
+}
+
+// Update the wait_on_escape_ms value in response to the fish_escape_delay_ms user variable being
+// set.
 pub fn update_wait_on_escape_ms(vars: &EnvStack) {
     let fish_escape_delay_ms = vars.get_unless_empty(L!("fish_escape_delay_ms"));
-    let is_empty = fish_escape_delay_ms.is_none();
-    let value = fish_escape_delay_ms
-        .map(|s| s.as_string().to_ffi())
-        .unwrap_or(L!("").to_ffi());
-    crate::ffi::update_wait_on_escape_ms_ffi(is_empty, &value);
+    let Some(fish_escape_delay_ms) = fish_escape_delay_ms else {
+        WAIT_ON_ESCAPE_MS.store(WAIT_ON_ESCAPE_DEFAULT, Ordering::Relaxed);
+        return;
+    };
+    let fish_escape_delay_ms = fish_escape_delay_ms.as_string();
+    match fish_wcstol(&fish_escape_delay_ms) {
+        Ok(val) if (10..5000).contains(&val) => {
+            WAIT_ON_ESCAPE_MS.store(val.try_into().unwrap(), Ordering::Relaxed);
+        }
+        _ => {
+            eprintln!(
```

Comment:
I found that we can't mix `print!` with `printf!` because `print!` uses a buffer that our
`printf!`/`wwrite_to_fd` doesn't use. The `println!` variants should be fine though.